src/wasm/wasm-ir-builder.cpp

```diff
@@ -282,18 +282,26 @@ Result<> IRBuilder::visitArrayNew(ArrayNew* curr) {
 }
 
 Result<> IRBuilder::visitBlockStart(Block* curr) {
-  scopeStack.push_back({{}, curr});
+  scopeStack.push_back(ScopeCtx::makeBlock(curr));
   return Ok{};
 }
 
-Result<> IRBuilder::visitEnd() {
-  if (scopeStack.empty() || !scopeStack.back().block) {
-    return Err{"unexpected end"};
+Result<> IRBuilder::visitIfStart(If* iff, Name label) {
+  auto cond = pop();
+  CHECK_ERR(cond);
+  iff->condition = *cond;
+  scopeStack.push_back(ScopeCtx::makeIf(iff, label));
+  return Ok{};
+}
+
+Result<Expression*> IRBuilder::finishScope(Block* block) {
+  if (scopeStack.empty() || scopeStack.back().isNone()) {
```

Comment:
So in the case where I want to outline a sequence of 4 instructions, let's say all of them i32.const, I would need to call makeBlockStart() first? To be on the safe side, should I always intend to create a block for the sequence to be outlined so that the scopeStack is not empty (or doesn't have a last value of none)?

Reply:
If you want those four i32.const instructions to be inside a block, then you would call either `visitBlockStart` or `makeBlock`, then `visit` on each of the i32.const instructions, then `visitEnd`. But if you're outlining the four instructions, then you may not know in advance whether you need a block, since it depends on whether you end up with one top-level instruction or multiple 🤔 

I think the best thing for me to do will be to add function scope contexts (which I'm realizing we need anyway for branches to top-level function scopes) so you can call `visitFunctionStart` at the beginning of a function (instead of `setFunction`) and `visitEnd` at the end of the function, which will then handle setting the function body for you.